File: udpclient.pl
#!/usr/bin/perl
use IO::Socket::INET;
use Time::HiRes qw(time);

my $mySocket=new IO::Socket::INET->new(LocalPort=>1234,
        Proto=>'udp');

my $threshold=5;
my $interval=1;
my $debug = 1;
my $print_jitter_interval = 5;

my $expcount=1; my $exptime=time();
my $lastime=time(); $lastpacket = 0;
while ( 1) {
        $mySocket->recv($text,1500);
        my $bytes= length($text);
        my ($instr,$var1, $var2, $var3, $var4) = split(/:/,$text);
        if ($instr eq 'DATA') {
                my ($count,$time,$payload, $cks) = ($var1, $var2, $var3, $var4);
                if ($cks ne chksum($payload)) {
                        print sprintf "checksum error [$cks] %s $count $time $payload\n" , chksum($payload);
                }
                if ($expcount != $count) {
                        my $dropped = $count-$expcount;
                        my $int = $dropped * $interval;
                        my $now = time();
                        if ($dropped > $threshold) {
                                print "$now Gap: $dropped packets, $int ms ($time, $count, $expcount)\n";
                                my $localtgap = ($now - $lasttime)*1000;
                                my $localpgap = $count - $lastpacket;
                                printf "Gap seen locally is $localtgap ($now - $lasttime), $localpgap ($count - $lastpacket)\n";
                        }
                } else {
                        $lasttime=time(); $lastpacket=$count;
                        my $gap = $time - $exptime;
                        if ($count % (1000*$print_jitter_interval/$interval) == 1 ){
                                print "jitter gap $gap ($count $time)\n";
                        }
                }
                $expcount = $count+1;
                $exptime = time()+$interval/1000;
        } elsif ($instr eq 'RSTTMR') {
                print "Setting Timer Interval to $interval\n" if $debug;
                $interval=$var1;
        } else {
                print "Unknown Instruction $instr\n";
        }
#       print "received packet $count\n";
}

exit;

sub chksum {
  my ($string) = @_;
  my $v = 0;
  $v ^= $_ for unpack 'C*', $string;
  my $ret = sprintf '%02X', $v;
  return $ret;
}

sub convms {
my $sec=shift;
return sprintf("%.3f",$sec*1000);
}

sub convspeed {
my $bytes=shift;
my $bits=$bytes*8;
my $units = "bps";
if ($bits < 1e3) {
        $units = "bps";
} elsif ($bits < 1e6) {
        $bits = $bits/1e3;
        $units = "kbps";
} elsif ($bits < 1e9) {
        $bits = $bits/1e6;
        $units = "Mbps";
} elsif ($bits < 1e12) {
        $bits = $bits/1e9;
        $units = "Gbps";
}
$bits=sprintf("%.1f",$bits);
return "$bits $units";
}

File: udp_client.py
import socket
import sys
import time

# Create a UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
server_address = ('0.0.0.0', 9090)
sock.bind(server_address)


threshold=5; logthreshlower = 0; logthreshupper = 10000;
interval=1; debug=0
print_jitter_interval=5

expcount=1;exptime=time.time()
lasttime=time.time();lastpacket=0;

def cksum(string):
  checksum=0
  for c in string:
     checksum ^=ord(c)
  return "%02X" % checksum

def convms(sec):
  return "%.3f" % (sec*1000)

def convspeed(bytes):
  bits=bytes*8
  units = "bps"
  if bits < 1e3:
     units = "bps"
  elif bits < 1e6:
     bits = bits/1e3
     units = "kbps"
  elif bits < 1e9:
     bits = bits/1e6
     units = "Mbps"
  elif bits < 1e12:
     bits = bits/1e9
     units = "Gbps"
  bits= "%.1f" % bits
  return "%s %s" % (str(bits), str(units))

while True:
   text, address = sock.recvfrom(1500)
   (instr,var1,var2,var3,var4) = text.split(':')
   if instr == 'DATA':
      (count,tme,payload,cks) = (int(var1),float(var2),var3,var4)
      if cks != cksum(payload):
         print "checksum error [%s] %s %s %s %s" % (str(cks), str(cksum(payload)),str(count), str(tme), str(payload))
      if expcount != count:
         dropped = count-expcount
         intt = dropped * interval
         now = time.time()
         if dropped > threshold:
            if debug:
               print "%s Gap: %s packets, %s ms (%s, %s %s)" % (str(now),str(dropped),str(intt),str(tme),str(count),str(expcount))
            localtgap = (now - lasttime)*1000
            localpgap = count - lastpacket
            if debug:
               print "Gap seen locally is %s (%s - %s), %s (%s - %s)" % (str(localtgap), str(now), str(lasttime), str(localpgap), str(count), str(lastpacket)
)
         if localtgap > logthreshlower and localtgap < logthreshupper:
            print localtgap
      else:
         lasttime=time.time()
         lastpacket=count
         gap=tme-exptime
         if count % (1000 *print_jitter_interval/interval) == 1:
            if debug:
               print "jitter gap %s (%s - %s)"% (str(gap),str(count),str(tme))
      expcount=count+1
      exptime=time.time()+interval/1000
   elif instr == 'RSTTMR':
      if debug:
         print "Settimg Time Interval to %s" % interval
      interval=float(var1)
   else:
      print "Unknown Instruction %s" % instr

